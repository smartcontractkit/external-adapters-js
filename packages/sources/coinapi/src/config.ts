import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'COINAPI'

export const DEFAULT_ENDPOINT = 'price'
export const DEFAULT_API_ENDPOINT = 'https://rest.coinapi.io/v1/'
export const DEFAULT_WS_API_ENDPOINT = 'wss://ws.coinapi.io/v1/'

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api = {
    ...config.api,
    baseURL: config.api.baseURL || DEFAULT_API_ENDPOINT,
    params: {
      apikey: config.apiKey,
    },
  }
  return config
}
