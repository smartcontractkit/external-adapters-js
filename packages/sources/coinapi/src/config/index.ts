import { Requester } from '@chainlink/ea-bootstrap'
import { DefaultConfig } from '@chainlink/ea-bootstrap'

export const NAME = 'COINAPI'

export const DEFAULT_ENDPOINT = 'crypto'
export const DEFAULT_BASE_URL = 'https://rest.coinapi.io/v1/'
export const DEFAULT_WS_API_ENDPOINT = 'wss://ws.coinapi.io/v1/'

export const makeConfig = (prefix?: string): DefaultConfig => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api = {
    ...config.api,
    baseURL: config.api.baseURL ?? DEFAULT_BASE_URL,
    params: {
      apikey: config.apiKey,
    },
  }
  config.defaultEndpoint = DEFAULT_ENDPOINT
  config.ws.baseWsURL = config.ws.baseWsURL || DEFAULT_WS_API_ENDPOINT
  return config
}
