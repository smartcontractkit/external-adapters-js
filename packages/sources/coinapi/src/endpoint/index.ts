export * as crypto from './crypto'
export * as assets from './assets'
