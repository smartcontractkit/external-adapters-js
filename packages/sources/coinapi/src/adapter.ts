import { AdapterError, Requester, Validator } from '@chainlink/ea-bootstrap'
import { Config, ExecuteFactory, ExecuteWithConfig, MakeWSHandler } from '@chainlink/types'
import { DEFAULT_ENDPOINT, DEFAULT_WS_API_ENDPOINT, makeConfig, NAME } from './config'
import { assets, price } from './endpoint'

const inputParams = {
  endpoint: false,
}

export const execute: ExecuteWithConfig<Config> = async (request, config) => {
  const validator = new Validator(request, inputParams)
  if (validator.error) throw validator.error

  Requester.logConfig(config)

  const jobRunID = validator.validated.id
  const endpoint = validator.validated.data.endpoint || DEFAULT_ENDPOINT

  switch (endpoint) {
    case price.NAME: {
      const quote = validator.validated.data.quote
      if (quote?.toUpperCase() === 'USD') return await assets.execute(request, config)
      return await price.execute(request, config)
    }
    case assets.NAME: {
      return await assets.execute(request, config)
    }
    default: {
      throw new AdapterError({
        jobRunID,
        message: `Endpoint ${endpoint} not supported.`,
        statusCode: 400,
      })
    }
  }
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request) => execute(request, config || makeConfig())
}

export const makeWSHandler =
  (config?: Config): MakeWSHandler =>
  () => {
    const defaultConfig = config || makeConfig()
    const getSubscription = (products: string[]) => ({
      type: 'hello',
      apikey: defaultConfig.apiKey,
      heartbeat: false,
      subscribe_data_type: ['exrate'],
      subscribe_filter_asset_id: products,
    })
    return {
      connection: {
        url: defaultConfig.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => {
        const validator = new Validator(input, price.customParams, {}, false)
        if (validator.error) return
        const base = (validator.overrideSymbol(NAME) as string).toLowerCase()
        const quote = validator.validated.data.quote.toLowerCase()
        return getSubscription([base, quote])
      },
      unsubscribe: () => undefined,
      subsFromMessage: (message) =>
        getSubscription([message.asset_id_base, message.asset_id_quote]),
      isError: () => false,
      filter: (message) => message?.type === 'exrate',
      toResponse: (message) => {
        const result = Requester.validateResultNumber(message, ['rate'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
