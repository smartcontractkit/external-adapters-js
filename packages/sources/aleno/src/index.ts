import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
