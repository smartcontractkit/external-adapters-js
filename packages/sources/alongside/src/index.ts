import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
