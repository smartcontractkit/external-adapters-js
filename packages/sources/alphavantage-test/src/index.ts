import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
