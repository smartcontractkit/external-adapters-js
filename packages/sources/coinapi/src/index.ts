import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig, NAME, makeWSHandler } from './config'

export = { NAME, makeExecute, makeConfig, ...expose(makeExecute(), makeWSHandler(makeConfig())) }
