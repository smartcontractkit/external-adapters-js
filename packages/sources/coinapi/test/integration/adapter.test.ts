import { AdapterRequest } from '@chainlink/types'
import { util } from '@chainlink/ea-bootstrap'
import http from 'http'
import nock from 'nock'
import request from 'supertest'
import { server as startServer } from '../../src/index'
import { mockCryptoEndpoint } from './cryptoFixtures'

describe('coinapi', () => {
  let server: http.Server
  const oldEnv: NodeJS.ProcessEnv = JSON.parse(JSON.stringify(process.env))
  const req = request('localhost:8080')

  beforeAll(async () => {
    process.env.CACHE_ENABLED = 'false'

    if (util.parseBool(process.env.RECORD)) {
      nock.recorder.rec()
    } else {
      process.env.API_KEY = 'mock-api-key'
    }
    server = await startServer()
  })
  afterAll((done) => {
    process.env = oldEnv
    if (util.parseBool(process.env.RECORD)) {
      nock.recorder.play()
    }

    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
    server.close(done)
  })

  describe('crypto endpoint', () => {
    describe('when sending well-formed request', () => {
      it('should reply with success', async () => {
        const cryptoRequest: AdapterRequest = {
          id: '1',
          data: {
            endpoint: 'crypto',
            base: 'ETH',
            quote: 'BTC',
          },
        }
        mockCryptoEndpoint()
        const response = await req
          .post('/')
          .send(cryptoRequest)
          .set('Accept', '*/*')
          .set('Content-Type', 'application/json')
          .expect('Content-Type', /json/)
          .expect(200)
        expect(response.body).toMatchSnapshot()
      })
    })
  })
})
