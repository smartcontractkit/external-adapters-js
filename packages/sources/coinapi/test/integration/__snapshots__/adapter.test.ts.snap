// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coinapi crypto endpoint when sending well-formed request should reply with success 1`] = `
Object {
  "data": Object {
    "result": 0.06262119731705901,
  },
  "jobRunID": "1",
  "result": 0.06262119731705901,
  "statusCode": 200,
}
`;
